existential-vs-universal-quantification/article.md

```diff
@@ -0,0 +1,334 @@
+
+First of all, you can now change the order of type variables.
+Let's imagine you want to define a function with 10 type variables (yes, there are real-life [examples](https://hackage.haskell.org/package/leancheck-0.9.10/docs/Test-LeanCheck-Utils-TypeBinding.html#v:-45--62--62--62--62--62--62--62--62--62--62--62--62-:) of such functions). And when you use it, you want to specify the type of the last argument explicitly.
+
+By default, the order of type variables that you assign types to is the same as the order in which they are introduced in the function definition.
+
+Without `forall`, you would write something like this: 
+
+```Haskell
+veryLongFunction :: a -> b -> c -> d -> e -> f -> g -> h -> i -> j
+veryLongFunction = ...
+
+func = veryLongFunction @_ @_ @_ @_ @_ @_ @_ @_ @_ @Integer ...
+```
+
+Quite long, right?
+However, this can be simplified with explicit `forall`:
+
+```Haskell
+veryLongFunction :: forall j a b c d e f g h i. a -> b -> c -> d -> e -> f -> g -> h -> i -> j
+veryLongFunction = ...
+
+func = veryLongFunction @Integer ...
+```
+
+Since `j` is the first type variable in the declaration of `veryLongFunction`, you can specify only it.
+This is useful not only in such big examples but also when your function requires any of its arguments to be specified on usage.
+
+### Supporting other extensions
+
+On its own, `ExplicitForAll` doesn't do a lot.
+
+But you frequently need it to enable other extensions. Some of them just won't work without the `forall` quantifier. :)
+
+The most common one is `ScopedTypeVariables`, which allows you to pass a type variable from an outer function to one defined in a where clause.
+
+The code below will not compile because the compiler thinks that the `a` in `example`'s signature and the `a` in `pair`'s are two different types.
+
+```haskell
+example :: a -> [a] -> [a]
+example x rest = pair ++ rest
+  where
+    pair :: [a]
+    pair = [x, x]
+```
+
+To fix it, you need to use the `ScopedTypeVariables` extension and also add a `forall` quantifier to the `example` function, since the extension won't work without it.
+
+```haskell
+example :: forall a. a -> [a] -> [a]
+```
+You might also notice how `forall` was useful with the `TypeApplication` extension.
+Other extensions that somehow benefit from the usage of `ExplicitForAll` are `LiberalTypeSynonyms`, `RankNTypes`, and many more.
+
+## Existential quantification
+
+As I said earlier, besides universally quantified variables, Haskell also supports existential types.
+They are also created with the `forall` keyword.
+
+They can be implemented this way because these two constructions – `(exists x. p x) -> q` and `forall x. (p x -> q)` – are equivalent in terms of first-order predicate logic.
+For a theoretical proof of this statement, you can check this [thread](https://stackoverflow.com/questions/10753073/whats-the-theoretical-basis-for-existential-types?rq=1).
+
+### Usage
```

Comment:
I think there should be two sections: existential quantification in datatypes, and existential quantification in functions instead of usage and heterogenous list.

Reply:
Yeah, that sounds better